File: 1250-longest-common-subsequence/longest-common-subsequence.java
class Solution {
    int[][] dp;
    public int longestCommonSubsequence(String text1, String text2) {
        int n1 = text1.length();
        int n2 = text2.length();
        dp = new int[n1][n2];
        for(int[] row : dp) Arrays.fill(row , -1);
        return solve(n1 - 1 , n2 - 1 , text1 , text2); 
    }

    private int solve(int i , int j , String text1 , String text2){
        if(i < 0 || j < 0){
            return 0;
        }

        if(dp[i][j] != -1) return dp[i][j];
        
        if(text1.charAt(i) == text2.charAt(j)){
            return dp[i][j] = 1 + solve(i-1 , j-1 , text1 , text2);
        }

        return dp[i][j] = Math.max(solve(i-1 , j , text1 , text2) , solve(i , j-1 , text1 , text2));
    }
}

File: 1250-longest-common-subsequence/longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m = len(text1)
        n = len(text2)

        prev = [0 for _ in range(n+1)] 
        curr = [0 for _ in range(n+1)] 

        for i in range(1,m+1):
            for j in range(1,n+1):
                if text1[i-1] == text2[j-1]:
                    curr[j] = 1 + prev[j-1]
                else:
                    curr[j] = max(
                        prev[j],
                        curr[j-1]
                    )
            
            prev = curr[:]
        
        return prev[n]
